src/cls/IPM/Repo/Oras/PackageService.cls

```diff
@@ -0,0 +1,463 @@
+Include %IPM.Common
+
+Class %IPM.Repo.Oras.PackageService Extends (%RegisteredObject, %IPM.Repo.IPackageService)
+{
+
+/// The base URL for this repo, e.g. https://irepo.intersystems.com
+/// The API endpoint is will be located at <Location>/v2/
+Property Location As %String;
+
+/// Optional extension for where packages get published to and pulled from
+/// i.e. the API endpoint for a given module will be at <Location>/v2/<Namespace>/<Module>
+Property Namespace As %String;
+
+Property Username As %String(MAXLEN = "");
+
+Property Password As %String(MAXLEN = "");
+
+Property Token As %String(MAXLEN = "");
+
+// ** PACKAGE SERVICE FUNCTIONS **
+
+Method GetModule(pModuleReference As %IPM.Storage.ModuleInfo, Output AsArchive As %Boolean = 0) As %Stream.Object
+{
+	Set name = pModuleReference.Name _ ":" _ pModuleReference.VersionString
+	Set status = ..Pull(..Location, name, ..Namespace, ..Username, ..Password, ..Token, .stream)
+	If $$$ISERR(status) {
+		$$$ThrowStatus(status)
+	}
+	#; module is pulled as a .tgz file
+	Set AsArchive = 1
+	Return stream
+}
+
+Method GetModuleManifest(pModuleReference As %IPM.Storage.ModuleInfo) As %Stream.Object
+{
+	Set name = pModuleReference.Name _ ":" _ pModuleReference.VersionString
+	Return ..GetModuleXMLPy(..Location, name)
+}
+
+Method HasModule(pModuleReference As %IPM.Storage.ModuleInfo) As %Boolean
+{
+	Set name = pModuleReference.Name _ ":" _ pModuleReference.VersionString
+	
+	#; Get ORAS client
+	Set client = ..GetClient(..Location, ..Username, ..Password, ..Token)
+
+	Return ..HasModulePy(..Location, name, ..Namespace, client)
+}
+
+Method ListModules(pSearchCriteria As %IPM.Repo.SearchCriteria) As %ListOfObjects(ELEMENTTYPE="%IPM.Storage.ModuleInfo")
+{
+	#; Get ORAS client
+	Set client = ..GetClient(..Location, ..Username, ..Password, ..Token)
+
+	#; Parse search criteria
+	Set name = $$$lcase(pSearchCriteria.Name)
+	Set tVersionExpression = pSearchCriteria.VersionExpression
+	Set tSC = ##class(%IPM.General.SemanticVersionExpression).FromString(pSearchCriteria.VersionExpression, .tVersionExpression)
+	$$$ThrowOnError(tSC)
+	
+	#; If namespace is defined, add it to the package URI being searched for
+	If name'="", ..Namespace'="" {
+		Set name = ..AppendURIs(..Namespace, name)
+	}
+
+	#; Get all modules
+	Set tList = ##class(%Library.ListOfObjects).%New()
+	Set request = ..GetHttpRequest()
+	
+	#; Make GET request
+	Set tSC=request.Get("/v2/_catalog")
```

Comment:
Is this the base URI for all ORAS registries? If not, this should be something that the user can configure when creating the repo

Reply:
Yep that's the standard REST endpoint for OCI registries. See [here](https://distribution.github.io/distribution/spec/api/).